### PR TITLE
feat(lsp): allow diagnostics to be disabled in a buffer

### PR DESCRIPTION
**Pull request by gpanders**
Add two new methods to allow diagnostics to be disabled (and re-enabled)
in the current buffer. When diagnostics are disabled they are simply not
displayed to the user, but they are still sent by the server and
processed by the client.

Disabling diagnotics can be helpful in a number of scenarios. For
example, if one is working on a buffer with an overwhelming amount of
diagnostic warnings it can be helpful to simply disable diagnostics
without disabling the LSP client entirely. This also allows users more
flexibility on when and how they may want diagnostic information to be
displayed. For example, some users may not want to display diagnostic
information until after the buffer is first written.

### CODE REVIEW

File: runtime/lua/vim/lsp/diagnostic.lua [hunk 1]
```diff
@@ -1282,6 +1287,43 @@ function M.set_loclist(opts)
     vim.cmd [[lopen]]
   end
 end
+
+--- Disable diagnostics for the given buffer
+--- @param bufnr (optional, number): Buffer handle, defaults to current
+-- Note that when diagnostics are disabled for a buffer, the server will still
+-- send diagnostic information and the client will still process it. The
+-- diagnostics are simply not displayed to the user.
+function M.disable(bufnr)
```
Comment (mjlbach) [inline runtime/lua/vim/lsp/diagnostic.lua:1296]: I think we should also optionally allow a client id to be passed in, and default to all clients attached to the buffer.

**Comment by mjlbach** [comment]
We probably want to also expose a method which allows the user to check if diagnostics are enabled on a buffer by a client, but that may be outside the scope of this pr.

**Review by mjlbach — APPROVED** [review]

**Review by mfussenegger — COMMENTED** [review]
Code itself looks okay to me.

@mjlbach do you think this will also address the other requirements that came up in regards to delaying the diagnostic display until insert mode is left?
Would be cool if we could solve this "diagnostics are too noisy" topic in a general or flexible way that satisfies most users. 
E.g. should we add some docs that this enable/disable could be used with autocmds InsertEnter/InsertLeave ?

**Comment by mjlbach** [comment]
@mfussenegger Maybe, I think closer to solving that would be something like https://github.com/neovim/neovim/pull/13748

The main issue is users want to decouple the display of diagnostics from the handler update. We could make it such that display could run on a luv timer, such that while in insert mode, diagnostics were never displayed, and after the user leaves insert mode, we only display diagnostics on the next tick. I think that would address the noisy diagnostics issues.

**Comment by gpanders** [comment]
> We probably want to also expose a method which allows the user to check if diagnostics are enabled on a buffer by a client, but that may be outside the scope of this pr.

Agreed. It's an easy addition so I went ahead and just updated the PR with a `is_enabled(bufnr, client_id)` method.

**Comment by mjlbach** [comment]
It would need to be in a follow-up PR :) I already merged.

**Comment by gpanders** [comment]
@mjlbach This needs one small bugfix.

The lines

```lua
  validate { bufnr = { bufnr, 'n' } }

  bufnr = (bufnr == 0 and api.nvim_get_current_buf()) or bufnr
```

in `M.clear` should be changed to

```lua
  bufnr = get_bufnr(bufnr)
```

(I added this change in my last force-push, but the PR had already been merged).

The only change here is that it allows the `bufnr` parameter to `vim.lsp.diagnostic.clear()` to be `nil`, which it is when `vim.lsp.diagnostic.disable()` is called without arguments.

Should I open a new PR or do you want to commit that directly?

**Comment by mjlbach** [comment]
Open a follow-up PR

**Comment by neovim-discourse** [comment]
This pull request has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/disable-diagnostics-by-default/917/3

**Pull request merged**